### ISSUE TITLE
Request response is not working with Kafka as transport

### ISSUE THREAD

**Issue by ukotdiya02** [zarusz/SlimMessageBus#46]
When I set kafka as provider Request Response is not working. 

I am using Sample.Simple.ConsoleApp
1. I changed provider as Kafka in  var provider = Provider.Kafka; in CreateMessageBus function
2. I am using local on premise Kafka Broker
2. As I am not using SSL in kafka so I had commented addssl function for producerconfig and conumerconfig
3. When I run the program I can see Request and response topic for Multiply in Kafka but getting timeout after this statement
var response = await bus.Send(new MultiplyRequest { Left = a, Right = b });

**Comment by zarusz** [comment]
hello @ukotdiya02 ,

I did check again and it does work fine against my test instance of cloud karafka.
SSL should not make much difference if your local broker does not require it.

Can you check if you created all these topics?

```cs
            if (provider == Provider.Kafka)
            {
                // Note: We are using the free plan of CloudKarafka to host the Kafka infrastructure. The free plan has a limit on topic you can get free and it requires these topic prefixes.
                topicForAddCommand = "4p5ma6io-test-ping";
                topicForMultiplyRequest = "4p5ma6io-multiply-request";
                topicForResponses = "4p5ma6io-responses";
            }
```
 
Also, what system and Kafka Broker version you're using?

**Comment by ukotdiya02** [comment]
Thanks for your response. Personally I would like to thank you and say you had done great job of abstracting Messagebus  and implementing RequestResponse on broker which are primarily for Pub-Sub. YOU library is easy to use and understand.

I am using Kafka version 2.4.
When I debug the sequence is
1. When ProduceToTransport function is called in MessageBUsBase it goes to KafkaMessageBus.ProduceToTransport and create request on  request  topic in Kafka
2. After that it comes to below line of SendInternal function in MessageBUsBase
var responseUntyped = await requestState.TaskCompletionSource.Task.ConfigureAwait(true);
After this line it post response on response topic in Kafka
3. Then I get  timeout error
However unlike memorybus OnResponseArrvied event is never fired when I use Kafka. Am I missing somethings as why it is not going into KafkaResponseProcessor class.

Regards
Uttam

On 05-May-2021, at 2:17 AM, Tomasz Maruszak ***@***.******@***.***>> wrote:


[EXTERNAL]
! Caution: This email originates from outside Secure Meters. Please be cautious before responding, opening attachments or clicking on links, unless you recognise the sender and know that the content is safe.


hello @ukotdiya02<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fukotdiya02&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Cca7ba8a15b9249a81c7f08d90f3ddf53%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637557580680127816%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=Xb2lbSrXPsop2qo32DOCfRksXyEZEqhm%2FUKLVJrHD2M%3D&reserved=0> ,

I did check again and it does work fine against my test instance of cloud karafka.
SSL should not make much difference if your local broker does not require it.

Can you check if you created all these topics?

            if (provider == Provider.Kafka)
            {
                // Note: We are using the free plan of CloudKarafka to host the Kafka infrastructure. The free plan has a limit on topic you can get free and it requires these topic prefixes.
                topicForAddCommand = "4p5ma6io-test-ping";
                topicForMultiplyRequest = "4p5ma6io-multiply-request";
                topicForResponses = "4p5ma6io-responses";
            }

Also, what system and Kafka Broker version you're using?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzarusz%2FSlimMessageBus%2Fissues%2F46%23issuecomment-832234725&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Cca7ba8a15b9249a81c7f08d90f3ddf53%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637557580680127816%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=UShqlN4VqhHjETrDXD5JBWNhOiWNFpPDU3%2BWS4YDDQs%3D&reserved=0>, or unsubscribe<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAT6CNFR7GJ4ZRVYAWXO3CATTMBMO7ANCNFSM44CFKQNQ&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Cca7ba8a15b9249a81c7f08d90f3ddf53%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637557580680137817%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=psUV2sF6nvte%2FtdI%2Ff%2BO6n4A9G1pApjlu%2FdYm66QVCI%3D&reserved=0>.

This e-mail and any files transmitted with it are for the sole use of the intended recipient(s) and may contain confidential and privileged information. If you are not the intended recipient, please contact the sender and destroy all copies and the original message. Secure Meters Limited or its subsidiaries (Secure), are unable to exercise control or guarantee the integrity of/over the content contained in e-mails and any views expressed in this message are those of the individual unless expressly stated with due authority of Secure. If you have any problems with the emails from Secure contact: ***@***.***

**Comment by zarusz** [comment]
@ukotdiya02 , thanks for the kind words on the SMB library, that always helps ;)

I've found a sample docker compose that runs a single instance kafka + zookeeper to reproduce your issue. After some testing, the request-response Kafka scenario of the ConsoleApp sample seems to run fine for me:

![image](https://user-images.githubusercontent.com/2655886/117212981-533f8380-adfb-11eb-8e92-f96ee679eb2e.png)

I've created a feature branch that adds the `docker-compose.yml` file under the `ConsoleApp` folder. Inside it already creates the necessary Kafka topics required for the sample to work.

That said, I did found one small issue where the first request would always time-out on a fresh install of kafka (consumer group does not exist yet). That was because the consumer would not yet start before the request had been already produced. To fix this I've adjusted the offset reset setting to:

```cs
                                ConsumerConfig = (config) =>
                                {
                                    config.AutoOffsetReset = AutoOffsetReset.Earliest;
                                    //AddSsl(config);
                                }
```

In short, you can run this yourself:

1. Checkout the feature branch
2. Install Docker Desktop
3. Open command line and navigate under `Samples/ConsoleApp`
4. `docker-compose up`
5. (wait some time until Kafka and Zookeeper startup) 
6. Run the sample Console App

I will likely incorporate that docker compose file into the mainline of the project if people wanted to run the infrastructure locally for debugging.

**Comment by ukotdiya02** [comment]
Thanks for the pain you had taken. Yes it is working fine with single node docker. However it is not with w node cluster. I will be looking into the issue. Just in case you know the reason when I debug the program
When I run with kafka docker image constructor of KafkaResponseProcesor get called but when I run it with my 3 mode kafka cluster  KafkaResponseProcesor does nougat called. I think this is reason for not working 3 node cluster mode. Can you give hint when KafkaResponseProcessor constructor will get called.   In Docker scenario it is getting called just after this statement

 var multiplyTask = Task.Factory.StartNew(() => MultiplyLoop(messageBus), CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);

Regards
Uttam

On 06-May-2021, at 3:23 AM, Tomasz Maruszak ***@***.******@***.***>> wrote:


[EXTERNAL]
! Caution: This email originates from outside Secure Meters. Please be cautious before responding, opening attachments or clicking on links, unless you recognise the sender and know that the content is safe.


@ukotdiya02<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fukotdiya02&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Ce13bc75f436f4ab37a6108d9101044c9%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637558484339649524%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=hr46E1RykY8h9FwHdF6E%2BX1ANop6m%2BaEQf88MXJfAtc%3D&reserved=0> , thanks for the kind words on the SMB library, that always helps ;)

I've found a sample docker compose that runs a single instance kafka + zookeeper to reproduce your issue. After some testing, the request-response Kafka scenario of the ConsoleApp sample seems to run fine for me:

[image]<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fuser-images.githubusercontent.com%2F2655886%2F117212981-533f8380-adfb-11eb-8e92-f96ee679eb2e.png&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Ce13bc75f436f4ab37a6108d9101044c9%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637558484339649524%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=L8jAqxOE2D7B3hjLAP9oHJL4Sb7fgoCXHhVtZM%2B5dSA%3D&reserved=0>

I've created a feature branch that adds the docker-compose.yml file under the ConsoleApp folder. Inside it already creates the necessary Kafka topics required for the sample to work.

That said, I did found one small issue where the first request would always time-out on a fresh install of kafka (consumer group does not exist yet). That was because the consumer would not yet start before the request had been already produced. To fix this I've adjusted the offset reset setting to:

                                ConsumerConfig = (config) =>
                                {
                                    config.AutoOffsetReset = AutoOffsetReset.Earliest;
                                    //AddSsl(config);
                                }

In short, you can run this yourself:

  1.  Checkout the feature branch
  2.  Install Docker Desktop
  3.  Open command line and navigate under Samples/ConsoleApp
  4.  docker-compose up
  5.  (wait some time until Kafka and Zookeeper startup)
  6.  Run the sample Console App

I will likely incorporate that docker compose file into the mainline of the project if people wanted to run the infrastructure locally for debugging.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fzarusz%2FSlimMessageBus%2Fissues%2F46%23issuecomment-833054851&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Ce13bc75f436f4ab37a6108d9101044c9%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637558484339659519%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=BfMt9ogLgZ1h31yAn067bdvwP7x1WKFcvmnwlPBvmHs%3D&reserved=0>, or unsubscribe<https://ind01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAT6CNFVXZH23S2FLAQBBCPLTMG46ZANCNFSM44CFKQNQ&data=04%7C01%7Cuttam.kotdiya%40securemeters.com%7Ce13bc75f436f4ab37a6108d9101044c9%7C7377900b5c3541b99a9c5e83c9932af1%7C0%7C0%7C637558484339659519%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=dWoEqxn9M9w8%2BGrLbdtHW1DEbl5lx0HTGrhOMeeNvo4%3D&reserved=0>.

This e-mail and any files transmitted with it are for the sole use of the intended recipient(s) and may contain confidential and privileged information. If you are not the intended recipient, please contact the sender and destroy all copies and the original message. Secure Meters Limited or its subsidiaries (Secure), are unable to exercise control or guarantee the integrity of/over the content contained in e-mails and any views expressed in this message are those of the individual unless expressly stated with due authority of Secure. If you have any problems with the emails from Secure contact: ***@***.***

**Comment by zarusz** [comment]
@ukotdiya02 my test cloud instance on https://cloudkarafka.com is a cluster of 3 nodes and it works fine. Also, on another commercial project this is working fine in a clustered environment.

Perhaps you have a higher topic-partition replication factor set than the number of nodes you have?
Recreate the topics with replication factor 1 for each topic and see if that helps.
I've noticed that if the partition cannot be replicated to a sufficient amount of nodes the topic goes into some invalid state.

Here is what I have on cloud karafka:

![image](https://user-images.githubusercontent.com/2655886/117585680-fd791d00-b113-11eb-9f08-ae2e5b9132aa.png)

Another thing to check is to view the logs on each node and zookeeper - it should indicate.

Lastly, it would help if you'd have a sample Kubernetes deployment I could run with your setup.
